File: backend/ai/graph.py
import os

from dotenv import load_dotenv
import os
from langchain_core.messages import AIMessageChunk
from dotenv import load_dotenv
from langgraph.graph import StateGraph
from langgraph.graph import START, END

from ai.state import State
from ai.configuration import Configuration
from ai.nodes import (
    preprocess,
    detect_biases,
    extract_claims,
    verify_claims_llm,
    verify_claims_web,
    aggregate_verdict,
    export_report
)
from typing import Any, List, Optional
from ai.observability import setup_tracing
from logger import logger
import json
import asyncio
load_dotenv()

if os.getenv("OPENAI_API_KEY") is None:
    raise ValueError("OPENAI_API_KEY is not set")

def continue_to_verify_claims(state: State):
    """
    Determine whether to continue to the claim verification step based on the configuration and extracted claims.
    If no claims are extracted, we end the graph. If the configuration specifies web verification, we proceed to that step.
    Otherwise, we proceed to LLM verification.
    """
    if len(state["claims"].items) == 0:
        return 'no_claims'
    if state["configuration"].claim_verification_source == "web":
        return 'web_verification'
    return 'llm_verification'

# Create our Agent Graph
builder = StateGraph(State, config_schema=Configuration)

# Define the nodes we will cycle between
builder.add_node("preprocess", preprocess)
builder.add_node("detect_biases", detect_biases)
builder.add_node("extract_claims", extract_claims)
builder.add_node("verify_claims_llm", verify_claims_llm)
builder.add_node("verify_claims_web", verify_claims_web)
builder.add_node("aggregate_verdict", aggregate_verdict, defer=True) # Defer aggregation until all claims are verified
builder.add_node("export_report", export_report)

# Build the graph structure
builder.add_edge(START, "preprocess")
# Parallelize the preprocessing step with the detection of biases and extraction|verification of claims
builder.add_edge("preprocess", "detect_biases")
builder.add_edge("preprocess", "extract_claims")
# Add a conditional edge to verify claims based on the configuration
builder.add_conditional_edges(
    "extract_claims",
    continue_to_verify_claims,
    {
        'no_claims': "aggregate_verdict",  # If no claims are extracted, verification is skipped
        'web_verification': "verify_claims_web",  # If web verification is needed
        'llm_verification': "verify_claims_llm"  # If LLM verification is needed
    }
)
# Finally fan out the two branches to the verdict aggregation step
builder.add_edge("detect_biases", "aggregate_verdict")
builder.add_edge("verify_claims_llm", "aggregate_verdict")
builder.add_edge("verify_claims_web", "aggregate_verdict")
builder.add_edge("aggregate_verdict", "export_report")
# Finalize the answer
builder.add_edge("export_report", END)

graph = builder.compile(name="myth-buster-ai")

async def run_graph(input: str, configuration: Configuration):
    """Run the graph with the given input and configuration."""
    response = await graph.ainvoke({
        "original_content": input,
        "configuration": configuration,
        "claims": [],
        "biases": [],
        "verified_claims": [],
    })
    return response

def get_graph_as_png():
    return graph.get_graph().draw_mermaid_png()


def format_sse(data: Any, event: Optional[str] = None) -> str:
    """
    Format data and event to string for SSE message
    """
    msg = ""
    if event:
        msg += f"event: {event}\n"
    msg += f"data: {json.dumps(data)}\n\n"
    return msg


# === Streaming runner ===
async def stream_graph_events(
input: str, configuration: Configuration

):
    """
    Invoke and stream events from the graph and yield them as SSE messages.
    Args:
        graph: The graph to stream events from.
        initial_state: The initial state to start the graph with.
        token_node: The node to stream tokens from.
        thread_id: The thread ID for the graph.
        observability: Whether to set up observability tracing.
        callbacks: Optional callbacks to be used in the graph.
    Yields:
        str: The formatted SSE message.
    """


    async for mode, chunk in graph.astream({
        "original_content": input,
        "configuration": configuration,
    }, stream_mode=["updates", "messages"]):

        if (mode == "messages" ):
            msg, metadata = chunk
            if (msg and isinstance(msg, AIMessageChunk) and metadata["langgraph_node"] == "aggregate_verdict"):
                yield format_sse(msg.content, event="completion-token")
        elif (mode == "updates"):
            for node_name, node_data in chunk.items():
                if not isinstance(node_data, dict):
                    continue
                node_events = node_data.get("events", [])
                
                yield format_sse({"events": node_events}, event="node-update")

    yield format_sse("", event="completed")

File: ai/graph.ts
import { Annotation, StateGraph, END, START } from "@langchain/langgraph";
import { GraphEvent, GrpahConfiguration } from "@/types/types";
import { preprocess } from "./nodes/preprocess";
import { extractClaims } from "./nodes/extractClaims";
import { verifyClaimsWeb } from "./nodes/verifyClaimsWeb";
import { verifyClaimsLlm } from "./nodes/verifyClaimsLlm";
import { detectBiases } from "./nodes/detectBiases";
import { reporter } from "./nodes/reporter";
import {
  ExtractedBias,
  ExtractedClaim,
  ExtractedSegment,
  ExtractedVerifiedClaim,
} from "./schema";

const arrayReducer = <T>(left: T[], right: T[]) => left.concat([...right]);
const defaultArray = <T>() => [] as T[];
const arrayOptions = {
  reducer: arrayReducer,
  default: defaultArray,
};

const GraphAnnotation = Annotation.Root({
  originalContent: Annotation<string>(), // Initial input from caller
  cleanedContent: Annotation<string>(), // Cleaned content after preprocessing
  configuration: Annotation<GrpahConfiguration>(),
  segments: Annotation<ExtractedSegment[]>(arrayOptions), // Segments of the content
  extractedBiases: Annotation<ExtractedBias[]>(arrayOptions), // Biases detected in the content
  extractedClaims: Annotation<ExtractedClaim[]>(arrayOptions), // Claims extracted from the content
  verifiedClaims: Annotation<ExtractedVerifiedClaim[]>(arrayOptions), // Model used for claims verification
  report: Annotation<string>(), // Final answer to the query
  events: Annotation<GraphEvent[]>(arrayOptions), // Event for ui display
});

export type GraphState = typeof GraphAnnotation.State;

function continueToClaimsVerification(state: GraphState) {
  // If no claims are extracted, we end the graph. If the configuration specifies web verification, we proceed to that step.
  // Otherwise, we proceed to LLM verification.
  if (state.extractedClaims.length === 0) {
    console.log("No claims extracted, bypassing claims verification.");

    return "no_claims";
  }
  if (state.configuration.claimVerificationSource === "web") {
    return "verifyClaimsWeb";
  }
  return "verifyClaimsLlm";
}

export function buildGraph() {
  // Define a new graph
  const workflow = new StateGraph(GraphAnnotation)
    .addNode("preprocess", preprocess)
    .addNode("extractClaims", extractClaims)
    .addNode("verifyClaimsWeb", verifyClaimsWeb)
    .addNode("verifyClaimsLlm", verifyClaimsLlm)
    .addNode("detectBiases", detectBiases)
    .addNode("reporter", reporter, { defer: true })

    .addEdge(START, "preprocess")
    .addEdge("preprocess", "extractClaims")
    .addEdge("preprocess", "detectBiases")

    .addConditionalEdges("extractClaims", continueToClaimsVerification, {
      verifyClaimsWeb: "verifyClaimsWeb",
      verifyClaimsLlm: "verifyClaimsLlm",
      no_claims: "reporter",
    })
    .addEdge("verifyClaimsLlm", "reporter")
    .addEdge("verifyClaimsWeb", "reporter")
    .addEdge("detectBiases", "reporter")
    .addEdge("reporter", END);

  return workflow.compile();
}
